iml-services/iml-action-runner/src/main.rs

```diff
@@ -49,13 +50,29 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     .map(|x| warp::reply::json(&x))
     .with(log);
 
-    let addr = unsafe { NetUnixListener::from_raw_fd(3) };
-
-    let listener = UnixListener::try_from(addr)?
-        .incoming()
-        .inspect_ok(|_| tracing::debug!("Client connected"));
-
-    tokio::spawn(warp::serve(routes).serve_incoming(listener));
+    match env::var("LISTEN_PID") {
```

Comment:
I suggest factoring out common part (`tokio::spawn`), unless borrow checker is going to get angry (but I think it shouldn't).

```rust
let listener = match env::var("LISTEN_PID") {
  Ok(_) => {
    …
    UnixListener::try_from(…)...
  }
  Err(_) => {
    …
    TcpListener::bind(&addr)…
  }
}
tokio::spawn(warp::serve(routes).serve_incoming(listener));
```

Reply:
@mkpankov I approached the problem this way initially and while each case does return a listener, they are two different types of listeners and since the match must return the same type this didn't work. I was hoping there would be an enum that I could use instead that would encompass both listener types but I couldn't find one. If you know of one let me know and I will be happy to change it.

Reply:
This is actually trickier than it looks because there are two different things that need to be resolved to the same type:

1. The outer `Stream` (can be mostly done with `impl`)
2. The inner stream (either a `TcpStream` or `UnixStream`).

Both `TcpStream` and `UnixStream` implement  `AsyncRead` + `AsyncWrite` from tokio (plus a bunch of other traits), so the key is basically erasing the concrete structs and getting the inner streams to unify to those types.

I suspect this will be a common enough task that we should abstract it to a util crate. I'll push a patch for that.